File: kotlin/0102-binary-tree-level-order-traversal.kt
class Solution {
    fun levelOrder(root: TreeNode?): List<List<Int>> {
        val res = mutableListOf<List<Int>>()

        if (root == null)
            return res

        val queue = LinkedList<TreeNode>()
        queue.add(root)

        while (!queue.isEmpty()) {
            val size = queue.size
            val level = mutableListOf<Int>()

            repeat(size) {
                val node = queue.poll()
                level.add(node.`val`)

                if (node.left != null)
                    queue.add(node.left)

                if (node.right != null)
                    queue.add(node.right)
            }

            res.add(level)
        }

        return res
    }
}

File: typescript/0102-binary-tree-level-order-traversal.ts
function levelOrder(root: TreeNode | null): number[][] {
    const levels: number[][] = [];

    function getHeight(node: TreeNode | null, height: number) {
        if (!node) return 0;

        if (node.left || node.right) {
            getHeight(node.left, height + 1);
            getHeight(node.right, height + 1);
        }

        if (levels[height]) {
            levels[height].push(node.val);
        } else {
            levels[height] = [node.val];
        }
    }

    getHeight(root, 0);

    return levels;
}